### PR TITLE
Issue #418: Abstract RollupDataWriter and QbeastStats 

### PR DESCRIPTION
**Pull request by JosepSampe**
## Description

This is a partial PR for https://github.com/Qbeast-io/qbeast-spark/issues/418 (related to https://github.com/Qbeast-io/qbeast-spark/issues/398), which abstracts the RollupDataWriter and creates a new interface for the QbeastStats

## Type of change

Refactoring

## Checklist:

Here is the list of things you should do before submitting this pull request:

- [x] New feature / bug fix has been committed following the [Contribution guide](https://github.com/Qbeast-io/qbeast-spark/blob/main/CONTRIBUTING.md).
- [ ] Add logging to the code following the [Contribution guide](https://github.com/Qbeast-io/qbeast-spark/blob/main/CONTRIBUTING.md).
- [ ] Add comments to the code (make it easier for the community!).
- [ ] Change the documentation.
- [ ] Add tests.
- [ ] Your branch is updated to the main branch (dependent changes have been merged).

### CODE REVIEW

File: src/main/scala/io/qbeast/core/model/QbeastFiles.scala [hunk 1]
```diff
@@ -47,7 +47,7 @@ case class IndexFile(
     modificationTime: Long,
     revisionId: RevisionID,
     blocks: IISeq[Block],
-    stats: Option[String] = null)
+    stats: Option[QbeastStats] = null)
```
Comment (cugni) [inline src/main/scala/io/qbeast/core/model/QbeastFiles.scala:50]: Set as default None, not null.

File: src/main/scala/io/qbeast/core/model/QbeastStats.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import com.fasterxml.jackson.databind.JsonDeserializer
+import com.fasterxml.jackson.databind.JsonMappingException
+import com.fasterxml.jackson.databind.JsonNode
+import com.fasterxml.jackson.databind.JsonSerializer
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.databind.SerializerProvider
+import com.fasterxml.jackson.module.scala.ClassTagExtensions
+import com.fasterxml.jackson.module.scala.DefaultScalaModule
+
+case class QbeastStats(
```
Comment (cugni) [inline src/main/scala/io/qbeast/core/model/QbeastStats.scala:32]: Converting this to JSON, should be Delta's responsibility, not Qbeast's. Not all format use JSON to store this info, and this should be something format-specifcit

File: src/main/scala/io/qbeast/spark/writer/RollupDataWriter.scala [hunk 3]
```diff
@@ -13,82 +13,54 @@
...
 import org.apache.spark.util.SerializableConfiguration
 
-import java.net.URI
 import scala.collection.mutable
 
 /**
  * Implementation of DataWriter that applies rollup to compact the files.
  */
-object DeltaRollupDataWriter extends DataWriter with DeltaStatsCollectionUtils {
+abstract class RollupDataWriter extends DataWriter {
```
Comment (cugni) [inline src/main/scala/io/qbeast/spark/writer/RollupDataWriter.scala:37]: use traits, not abstract classes

Comment (JosepSampe) [reply src/main/scala/io/qbeast/spark/writer/RollupDataWriter.scala]: Changed

File: src/main/scala/io/qbeast/core/model/QbeastFiles.scala (hunk 1 above)
Comment (JosepSampe) [reply src/main/scala/io/qbeast/core/model/QbeastFiles.scala]: Fixed

File: src/main/scala/io/qbeast/core/model/QbeastStats.scala (hunk 2 above)
Comment (JosepSampe) [reply src/main/scala/io/qbeast/core/model/QbeastStats.scala:18]: Moved

**Review by cugni — APPROVED** [review]
Great job!

**Pull request merged**